### ISSUE TITLE
Can't use autodiff::jacobian in functions of different independent variables

### ISSUE THREAD

**Issue by Q-Minh** [autodiff/autodiff#192]
Hi, 

I am currently working on elasticity problems, in which the function to find is a mapping `x(X)` from some domain `X in R3` to some codomain `x in R3`. We approximate this mapping by the interpolation `x(X) = x1 * phi1(X) + x2 * phi2(X) + x3 * phi3(X) + x4 * phi4(X)`, where the `xi`s represent positions in the codomain `x` and the `phi`s are some basis functions (polynomials) acting on positions in the domain `X`.  Working on elasticity problems will require differentiating with respect to `X` to compute a quantity `F=dx/dX`, the jacobian of the mapping `x(X)`. However, we then use this quantity `F` to compute a strain quantity `E = 1/2 (F.transpose() * F - I)`, and then use this quantity `E` to compute an energy `Psi(X,x) = mu*(E:E) + 1/2 lambda * tr(E)^2`. Up until now, we have only needed derivatives with respect to the domain `X`. But then I also need to compute derivatives of this energy with respect to `x`, which will give me "elastic forces" (not exactly) at the nodes (omitting integration for simplicity) as `(dPsi/dx1, dPsi/dx2, dPsi/dx3, dPsi/dx4)`. 

My problem is that when I use the function `autodiff::jacobian(Fun f, Wrt wrt, At at)` to compute `F = dx/dX`, the energy derivatives are zeroed, which is an error. But when I instead myself implement the jacobian as a sum of outer products of `xi`s with `phis`, I get the correct energy derivatives (elastic forces). I would like to know why this is the case, as I would find it much cleaner to be able to use the `autodiff::jacobian` function to compute jacobians instead of gradient outer products using `autodiff::gradient`.

Here is a reproducible example program for my particular scenario:
```
#include <Eigen/Geometry>
#include <autodiff/forward/dual.hpp>
#include <autodiff/forward/dual/eigen.hpp>
#include <iostream>

autodiff::Vector4dual polynomial3d_1st_order(autodiff::Vector3dual X)
{
    return autodiff::Vector4dual(1.0, X.x(), X.y(), X.z());
}

struct basis_function_op_t
{
    autodiff::dual operator()(autodiff::Vector3dual X) const
    {
        autodiff::Vector4dual p1 = polynomial3d_1st_order(X);
        return p1.dot(a);
    }

    autodiff::Vector4dual a;
};

struct interpolation_op_t
{
    interpolation_op_t(
        autodiff::Vector3dual const& x1,
        autodiff::Vector3dual const& x2,
        autodiff::Vector3dual const& x3,
        autodiff::Vector3dual const& x4,
        basis_function_op_t const& phi1,
        basis_function_op_t const& phi2,
        basis_function_op_t const& phi3,
        basis_function_op_t const& phi4)
        : phi1(phi1), phi2(phi2), phi3(phi3), phi4(phi4)
    {
    }

    autodiff::Vector3dual operator()(
        autodiff::Vector3dual X,
        autodiff::Vector3dual x1,
        autodiff::Vector3dual x2,
        autodiff::Vector3dual x3,
        autodiff::Vector3dual x4) const
    {
        return x1 * phi1(X) + x2 * phi2(X) + x3 * phi3(X) + x4 * phi4(X);
    }
    basis_function_op_t phi1, phi2, phi3, phi4;
};

struct deformation_gradient_op_t
{
    deformation_gradient_op_t(interpolation_op_t const& interpolate) : interpolate_op(interpolate)
    {
    }

    autodiff::Matrix3dual operator()(
        autodiff::Vector3dual X,
        autodiff::Vector3dual x1,
        autodiff::Vector3dual x2,
        autodiff::Vector3dual x3,
        autodiff::Vector3dual x4,
        Eigen::Vector3dual& u) const
    {
        using autodiff::at;
        using autodiff::gradient;
        using autodiff::jacobian;
        using autodiff::wrt;

        autodiff::dual phi1, phi2, phi3, phi4;
        autodiff::Vector3dual gradphi1 = gradient(interpolate_op.phi1, wrt(X), at(X), phi1);
        autodiff::Vector3dual gradphi2 = gradient(interpolate_op.phi2, wrt(X), at(X), phi2);
        autodiff::Vector3dual gradphi3 = gradient(interpolate_op.phi3, wrt(X), at(X), phi3);
        autodiff::Vector3dual gradphi4 = gradient(interpolate_op.phi4, wrt(X), at(X), phi4);

        u                       = x1 * phi1 + x2 * phi2 + x3 * phi3 + x4 * phi4;
        // IMPORTANT:
        // Computing the jacobian manually as outer products xi * gradphi works! 
        // Final gradients dPsi/dxi are correctly computed later on.
        autodiff::Matrix3dual F = x1 * gradphi1.transpose() + x2 * gradphi2.transpose() +
                                  x3 * gradphi3.transpose() + x4 * gradphi4.transpose();

        // IMPORTANT: 
        // Here, if I use jacobian, then I cannot compute dPsi/dxi correctly later on, 
        // because the gradients are zeroed out.
        // autodiff::Matrix3dual F = jacobian(interpolate_op, wrt(X), at(X, x1, x2, x3, x4));
        return F;
    }

    interpolation_op_t interpolate_op;
};

struct strain_op_t
{
    strain_op_t(deformation_gradient_op_t const& deformation_gradient)
        : deformation_gradient_op(deformation_gradient)
    {
    }

    autodiff::Matrix3dual operator()(
        autodiff::Vector3dual X,
        autodiff::Vector3dual x1,
        autodiff::Vector3dual x2,
        autodiff::Vector3dual x3,
        autodiff::Vector3dual x4,
        autodiff::Vector3dual& u,
        autodiff::Matrix3dual& F) const
    {
        autodiff::Matrix3dual I = autodiff::Matrix3dual::Identity();
        F                       = deformation_gradient_op(X, x1, x2, x3, x4, u);
        autodiff::Matrix3dual E = 0.5 * (F.transpose() * F - I);
        return E;
    }

    deformation_gradient_op_t deformation_gradient_op;
};

struct strain_energy_density_op_t
{
    strain_energy_density_op_t(strain_op_t const& strain, double mu, double lambda)
        : strain_op(strain), mu(mu), lambda(lambda)
    {
    }

    autodiff::dual operator()(
        autodiff::Vector3dual X,
        autodiff::Vector3dual x1,
        autodiff::Vector3dual x2,
        autodiff::Vector3dual x3,
        autodiff::Vector3dual x4,
        autodiff::Vector3dual& u,
        autodiff::Matrix3dual& F,
        autodiff::Matrix3dual& E) const
    {
        E                    = strain_op(X, x1, x2, x3, x4, u, F);
        autodiff::dual trace = E.trace();
        auto tr2             = trace * trace;
        autodiff::dual EdotE = (E.array() * E.array()).sum(); // contraction
        return mu * EdotE + 0.5 * lambda * tr2;
    }

    strain_op_t strain_op;
    double mu, lambda;
};

int main()
{
    autodiff::Vector3dual X1, X2, X3, X4;
    X1 << 0., 0., 0.;
    X2 << 1., 0., 0.;
    X3 << 0., 1., 0.;
    X4 << 0., 0., 1.;

    autodiff::Matrix3dual S;
    S = 2. * autodiff::Matrix3dual::Identity();

    autodiff::Vector3dual x1 = S * X1;
    autodiff::Vector3dual x2 = S * X2;
    autodiff::Vector3dual x3 = S * X3;
    autodiff::Vector3dual x4 = S * X4;

    autodiff::Matrix4dual A;
    A.col(0) = polynomial3d_1st_order(X1);
    A.col(1) = polynomial3d_1st_order(X2);
    A.col(2) = polynomial3d_1st_order(X3);
    A.col(3) = polynomial3d_1st_order(X4);

    autodiff::Matrix4dual Ainv;
    Ainv = A.inverse();

    basis_function_op_t phi1, phi2, phi3, phi4;
    phi1.a = Ainv.row(0);
    phi2.a = Ainv.row(1);
    phi3.a = Ainv.row(2);
    phi4.a = Ainv.row(3);

    double young_modulus = 1e6;
    double poisson_ratio = 0.35;
    double mu            = (young_modulus) / (2. * (1 + poisson_ratio));
    double lambda =
        (young_modulus * poisson_ratio) / ((1 + poisson_ratio) * (1 - 2 * poisson_ratio));

    interpolation_op_t interpolate_op(x1, x2, x3, x4, phi1, phi2, phi3, phi4);
    deformation_gradient_op_t deformation_gradient_op(interpolate_op);
    strain_op_t strain_op(deformation_gradient_op);
    strain_energy_density_op_t strain_energy_density_op(strain_op, mu, lambda);

    autodiff::Vector3dual X;
    X << 0., 1., 0.;

    autodiff::Vector3dual x;
    autodiff::Matrix3dual F, E;
    autodiff::dual Psi;
    autodiff::Vector3dual f1 = autodiff::gradient(
        strain_energy_density_op,
        autodiff::wrt(x1),
        autodiff::at(X, x1, x2, x3, x4, x, F, E),
        Psi);
    autodiff::Vector3dual f2 = autodiff::gradient(
        strain_energy_density_op,
        autodiff::wrt(x2),
        autodiff::at(X, x1, x2, x3, x4, x, F, E),
        Psi);
    autodiff::Vector3dual f3 = autodiff::gradient(
        strain_energy_density_op,
        autodiff::wrt(x3),
        autodiff::at(X, x1, x2, x3, x4, x, F, E),
        Psi);
    autodiff::Vector3dual f4 = autodiff::gradient(
        strain_energy_density_op,
        autodiff::wrt(x4),
        autodiff::at(X, x1, x2, x3, x4, x, F, E),
        Psi);

    std::cout << "x:\n" << x << "\n";
    std::cout << "F:\n" << F << "\n";
    std::cout << "E:\n" << E << "\n";
    std::cout << "Psi:\n" << Psi << "\n";
    std::cout << "f1:\n" << f1 << "\n";
    std::cout << "f2:\n" << f2 << "\n";
    std::cout << "f3:\n" << f3 << "\n";
    std::cout << "f4:\n" << f4 << "\n";

    return 0;
}
```

The output of the program when I manually compute the jacobian `F` is:
```
x:
0
2
0
F:
2 0 0
0 2 0
0 0 2
E:
1.5   0   0
  0 1.5   0
  0   0 1.5
Psi:
1.125e+07
f1:
-1e+07
-1e+07
-1e+07
f2:
1e+07
    0
    0
f3:
    0
1e+07
    0
f4:
    0
    0
1e+07
```

while the output of the program when I use `autodiff::jacobian` is:
```
x:
0
2
0
F:
2 0 0
0 2 0
0 0 2
E:
1.5   0   0
  0 1.5   0
  0   0 1.5
Psi:
1.125e+07
f1:
0
0
0
f2:
0
0
0
f3:
0
0
0
f4:
0
0
0
```

Notice how the resulting forces `f1, f2, f3, f4` differ in both cases. I have marked the sections in my example program where I encounter my problem with the comment `// IMPORTANT: [...]`. Specifically, the problem arises in the call operator of the `deformation_gradient_op_t` class.

Thank you for this amazing library and I hope to find an answer to my problem with your help.

**Comment by allanleal** [comment]
Hi @Q-Minh ,

Would you kindly produce a simplified reproducible example? I'm currently very busy with other projects and I need something simpler to try here so that I don't spend too much time trying to identify the underlying issue.

Best regards,
Allan

**Comment by Q-Minh** [comment]
Sure, sorry for the long code sample.

Here is a simplified version without all the function compositions. It's not too important to look at how I build the basis functions (I believe). The main problem is in the function `psi`, in which I've left comments.

```
#include <Eigen/Geometry>
#include <array>
#include <autodiff/forward/dual.hpp>
#include <autodiff/forward/dual/eigen.hpp>
#include <iostream>

autodiff::Vector4dual polynomial3d_1st_order(autodiff::Vector3dual X)
{
    return autodiff::Vector4dual(1.0, X.x(), X.y(), X.z());
}

struct linear_hat_basis_function_op_t
{
    autodiff::dual operator()(autodiff::Vector3dual X) const
    {
        autodiff::Vector4dual p1 = polynomial3d_1st_order(X);
        return p1.dot(a);
    }

    autodiff::Vector4dual a;
};

std::tuple<
    linear_hat_basis_function_op_t,
    linear_hat_basis_function_op_t,
    linear_hat_basis_function_op_t,
    linear_hat_basis_function_op_t>
get_basis_functions(
    autodiff::Vector3dual X1,
    autodiff::Vector3dual X2,
    autodiff::Vector3dual X3,
    autodiff::Vector3dual X4,
    autodiff::Vector3dual x1,
    autodiff::Vector3dual x2,
    autodiff::Vector3dual x3,
    autodiff::Vector3dual x4)
{
    autodiff::Matrix4dual A;
    A.col(0) = polynomial3d_1st_order(X1);
    A.col(1) = polynomial3d_1st_order(X2);
    A.col(2) = polynomial3d_1st_order(X3);
    A.col(3) = polynomial3d_1st_order(X4);

    autodiff::Matrix4dual Ainv;
    Ainv = A.inverse();

    linear_hat_basis_function_op_t phi1, phi2, phi3, phi4;
    phi1.a = Ainv.row(0);
    phi2.a = Ainv.row(1);
    phi3.a = Ainv.row(2);
    phi4.a = Ainv.row(3);

    return {phi1, phi2, phi3, phi4};
}

autodiff::Vector3dual interpolate(
    autodiff::Vector3dual X,
    autodiff::Vector3dual x1,
    autodiff::Vector3dual x2,
    autodiff::Vector3dual x3,
    autodiff::Vector3dual x4,
    linear_hat_basis_function_op_t const& phi1,
    linear_hat_basis_function_op_t const& phi2,
    linear_hat_basis_function_op_t const& phi3,
    linear_hat_basis_function_op_t const& phi4)
{
    autodiff::Vector3dual u = x1 * phi1(X) + x2 * phi2(X) + x3 * phi3(X) + x4 * phi4(X);
    return u;
}

autodiff::dual
psi(autodiff::Vector3dual X,
    autodiff::Vector3dual x1,
    autodiff::Vector3dual x2,
    autodiff::Vector3dual x3,
    autodiff::Vector3dual x4,
    linear_hat_basis_function_op_t const& phi1,
    linear_hat_basis_function_op_t const& phi2,
    linear_hat_basis_function_op_t const& phi3,
    linear_hat_basis_function_op_t const& phi4,
    double mu,
    double lambda)
{
    using autodiff::at;
    using autodiff::gradient;
    using autodiff::jacobian;
    using autodiff::wrt;

    autodiff::Vector3dual gradphi1 = gradient(phi1, wrt(X), at(X));
    autodiff::Vector3dual gradphi2 = gradient(phi2, wrt(X), at(X));
    autodiff::Vector3dual gradphi3 = gradient(phi3, wrt(X), at(X));
    autodiff::Vector3dual gradphi4 = gradient(phi4, wrt(X), at(X));

    // IMPORTANT:
    // COMPUTING JACOBIAN MANUALLY WITH autodiff::gradient.
    // WORKS FOR COMPUTING dPsi/dxi LATER ON.
    // COMMENT THIS SECTION TO DEBUG THE PROBLEM.
    autodiff::Matrix3dual F = x1 * gradphi1.transpose() + x2 * gradphi2.transpose() +
                              x3 * gradphi3.transpose() + x4 * gradphi4.transpose();

    // IMPORTANT:
    // USING autodiff::jacobian.
    // DOESN'T WORK FOR COMPUTING dPsi/dxi LATER ON.
    // UNCOMMENT THIS SECTION TO DEBUG THIS PART.
    // autodiff::Matrix3dual F =
    //    jacobian(interpolate, wrt(X), at(X, x1, x2, x3, x4, phi1, phi2, phi3, phi4));

    autodiff::Matrix3dual I = autodiff::Matrix3dual::Identity();
    autodiff::Matrix3dual E = 0.5 * (F.transpose() * F - I);
    autodiff::dual trace    = E.trace();
    auto tr2                = trace * trace;
    autodiff::dual EdotE    = (E.array() * E.array()).sum(); // contraction
    autodiff::dual density  = mu * EdotE + 0.5 * lambda * tr2;
    return density;
}

int main()
{
    autodiff::Vector3dual X1, X2, X3, X4;
    X1 << 0., 0., 0.;
    X2 << 1., 0., 0.;
    X3 << 0., 1., 0.;
    X4 << 0., 0., 1.;

    autodiff::Matrix3dual S;
    S = 2. * autodiff::Matrix3dual::Identity();

    autodiff::Vector3dual x1 = S * X1;
    autodiff::Vector3dual x2 = S * X2;
    autodiff::Vector3dual x3 = S * X3;
    autodiff::Vector3dual x4 = S * X4;

    auto const [phi1, phi2, phi3, phi4] = get_basis_functions(X1, X2, X3, X4, x1, x2, x3, x4);

    double young_modulus = 1e6;
    double poisson_ratio = 0.35;
    double mu            = (young_modulus) / (2. * (1 + poisson_ratio));
    double lambda =
        (young_modulus * poisson_ratio) / ((1 + poisson_ratio) * (1 - 2 * poisson_ratio));

    autodiff::Vector3dual X;
    X << 0., 1., 0.;

    using autodiff::at;
    using autodiff::gradient;
    using autodiff::wrt;

    // IMPORTANT: 
    // Want to compute dPsi/dxi here to find the forces f1, f2, f3, f4
    autodiff::dual Psi;
    autodiff::Vector3dual f1 =
        gradient(psi, wrt(x1), at(X, x1, x2, x3, x4, phi1, phi2, phi3, phi4, mu, lambda), Psi);
    autodiff::Vector3dual f2 =
        gradient(psi, wrt(x2), at(X, x1, x2, x3, x4, phi1, phi2, phi3, phi4, mu, lambda), Psi);
    autodiff::Vector3dual f3 =
        gradient(psi, wrt(x3), at(X, x1, x2, x3, x4, phi1, phi2, phi3, phi4, mu, lambda), Psi);
    autodiff::Vector3dual f4 =
        gradient(psi, wrt(x4), at(X, x1, x2, x3, x4, phi1, phi2, phi3, phi4, mu, lambda), Psi);

    std::cout << "Psi:\n" << Psi << "\n";
    std::cout << "f1:\n" << f1 << "\n";
    std::cout << "f2:\n" << f2 << "\n";
    std::cout << "f3:\n" << f3 << "\n";
    std::cout << "f4:\n" << f4 << "\n";

    return 0;
}
```

Expected output (I get this when I don't use `autodiff::jacobian`):
```
Psi:
1.125e+07
f1:
1e+07
1e+07
1e+07
f2:
-1e+07
    -0
    -0
f3:
    -0
-1e+07
    -0
f4:
    -0
    -0
-1e+07
```

Wrong output (when using `autodiff::jacobian`):
```
Psi:
1.125e+07
f1:
-0
-0
-0
f2:
-0
-0
-0
f3:
-0
-0
-0
f4:
-0
-0
-0
```

Please let me know if I should simplify more.

Best regards,
Minh

**Comment by Q-Minh** [comment]
Actually, I have to apologize again. It hadn't occurred to me that you probably don't really care much about the validity of the formulas. I have a super simple example with garbage formulas which illustrates the discrepancy between using the `autodiff::jacobian` method vs computing the jacobian manually. Here is a much simpler example:

```
#include <Eigen/Geometry>
#include <array>
#include <autodiff/forward/dual.hpp>
#include <autodiff/forward/dual/eigen.hpp>
#include <iostream>

autodiff::Vector4dual polynomial3d_1st_order(autodiff::Vector3dual X)
{
    return autodiff::Vector4dual(1.0, X.x(), X.y(), X.z());
}

struct linear_hat_basis_function_op_t
{
    autodiff::dual operator()(autodiff::Vector3dual X) const
    {
        autodiff::Vector4dual p1 = polynomial3d_1st_order(X);
        return p1.dot(a);
    }

    autodiff::Vector4dual a;
};

autodiff::Vector3dual interpolate(
    autodiff::Vector3dual X,
    autodiff::Vector3dual x,
    linear_hat_basis_function_op_t const& phi)
{
    autodiff::Vector3dual u = x * phi(X);
    return u;
}

autodiff::dual
energy(autodiff::Vector3dual X, autodiff::Vector3dual x, linear_hat_basis_function_op_t const& phi)
{
    using autodiff::at;
    using autodiff::gradient;
    using autodiff::jacobian;
    using autodiff::wrt;

    autodiff::Vector3dual gradphi = gradient(phi, wrt(X), at(X));

    // IMPORTANT:
    // WORKS FOR COMPUTING dPsi/dxi LATER ON.
    // COMMENT THIS SECTION TO DEBUG THE PROBLEM.
    autodiff::Matrix3dual Jacobian = x * gradphi.transpose();

    // IMPORTANT:
    // DOESN'T WORK FOR COMPUTING dPsi/dxi LATER ON
    // UNCOMMENT THIS SECTION TO DEBUG THIS PART.
    //autodiff::Matrix3dual Jacobian = jacobian(interpolate, wrt(X), at(X, x, phi));

    autodiff::dual Energy = Jacobian.squaredNorm();
    return Energy;
}

int main()
{
    linear_hat_basis_function_op_t phi{};
    phi.a = autodiff::Vector4dual(1., 1., 1., 1.);
    autodiff::Vector3dual X;
    X << 1., 0., 0.;
    autodiff::Vector3dual x;
    x << 2., 0., 0.;

    using autodiff::at;
    using autodiff::gradient;
    using autodiff::wrt;

    autodiff::dual Psi;
    autodiff::Vector3dual f = -gradient(energy, wrt(x), at(X, x, phi), Psi);

    std::cout << "Psi:\n" << Psi << "\n";
    std::cout << "f:\n" << f << "\n";

    return 0;
}
```

Expected output (manually computing the jacobian):
```
Psi:
12
f:
-12
 -0
 -0
```

Wrong output (using `autodiff::jacobian`):
```
Psi:
24
f:
-0
-0
-0
```

I've noticed that even the resulting jacobian is not equivalent when computing manually vs when computing with `autodiff::jacobian`.

Best regards, 
Minh

**Comment by allanleal** [comment]
Hi Minh,

The issue is that you are "consuming" the first-order `dual` numbers inside the `energy` function to compute `gradphi` and `Jacobian`. You should either compute these manually, or increase the order of the dual number to second order (`dual2nd`).

If performance is critical to you, I would not go 2nd order and compute these derivatives inside the `energy` function manually, using first-order `dual` (which is equivalent to `dual1st`). Of course, if this is a viable option. 

Hope this helps and clarify. Let me know! 

Best regards,
Allan

**Comment by Q-Minh** [comment]
Hi Allan,

The thing is, when I am computing `gradphi`, and then computing the outer product `autodiff::Matrix3dual Jacobian = x * gradphi.transpose()`, that **is the alternative** to computing `autodiff::Matrix3dual Jacobian = autodiff::jacobian(...)`. So I don't use `autodiff::gradient` and `autodiff::jacobian` together, it is one or the other. More clearly, What works is:
```
autodiff::Vector3dual gradphi = gradient(phi, wrt(X), at(X));
autodiff::Matrix3dual Jacobian = x * gradphi.transpose();
```
which is my definition of "computing the jacobian manually", and what doesn't work is:
```
autodiff::Matrix3dual Jacobian = jacobian(interpolate, wrt(X), at(X, x, phi));
```
So when I compute the jacobian manually in the energy function, by using `autodiff::gradient` w.r.t. `X`, it is able to carry out a derivative w.r.t. `x` later on. So the "consuming" part seems to only consume the derivatives w.r.t. `X`, but derivatives w.r.t. `x` do not vanish, which is good, because I need them later on to compute energy derivatives w.r.t. `x`. In opposition, when using `autodiff::jacobian` in the energy function to let autodiff take care of jacobian computation, the "consuming" part seems to consume both derivatives w.r.t. `X` and w.r.t. `x`, such that later on, I will not be able to compute energy derivatives w.r.t. `x` (the energy gradients vanish), but I need those gradients. So right now, I am using the "computing the jacobian manually" implementation, but I would like to be able to use `autodiff::jacobian` instead.

For the second point about using `dual2nd`, I believe `dual2nd` would carry 2nd order derivatives, which is not needed in my use case.

I hope I am being clear, thank you for taking the time to look into this issue and maintaining this repo.

**Comment by allanleal** [comment]
Hi Minh,

I have not been able to test your code yet. But I do see above the following:

1. `autodiff::gradient` is executed first here `autodiff::Vector3dual f = -gradient(energy, wrt(x), at(X, x, phi), Psi);`
2. `autodiff::gradient` is executed again inside the `energy` function here `autodiff::Vector3dual gradphi = gradient(phi, wrt(X), at(X));`

That's why I'm saying you would need 2nd order derivatives to be able to compute the first order derivative of `energy` wrt to `X`. The output of `gradient` when using 1st order dual is `VectorXd`. Try using 2nd order and see if this resolves.

I may be able to check this more closely next week, but please create a super minimal example.

**Comment by Q-Minh** [comment]
Hi Allan,

The derivatives are with respect to different variables, that is why I only need first order derivatives. There is one derivative w.r.t. `X` as an uppercase `X` and another derivative w.r.t. `x` as a lowercase `x`. I should change the lowercase `x` for `u` instead or something to make the example more clear. So now, we have an energy function `E(X,u)` and we want its gradient w.r.t. to `u`. However, inside the energy function `E(X,u)`, we will have to compute a jacobian w.r.t. `X`. When computing this jacobian using `autodiff::jacobian`, I cannot compute the gradient of `E(X,u)` w.r.t. `u` anymore. However, when I compute this jacobian manually using `autodiff::gradient` calls on the function `phi`, I am able to correctly recover the gradient of `E(X,u)` later on using `autodiff::gradient` calls on the function `E(X,u)`.

As for the super minimal example, I will do what I can, but I'm afraid I can't go that much simpler than the latest sample I've shared on this thread while preserving the nature of the problem. Here is the best I can do to illustrate the problem:
```
#include <Eigen/Geometry>
#include <array>
#include <autodiff/forward/dual.hpp>
#include <autodiff/forward/dual/eigen.hpp>
#include <iostream>

autodiff::Vector4dual polynomial3d_1st_order(autodiff::Vector3dual X)
{
    return autodiff::Vector4dual(1.0, X.x(), X.y(), X.z());
}

autodiff::dual phi(autodiff::Vector3dual X, autodiff::Vector4dual a)
{
    autodiff::Vector4dual p1 = polynomial3d_1st_order(X);
    return p1.dot(a);
}

autodiff::Vector3dual
interpolate(autodiff::Vector4dual a, autodiff::Vector3dual X, autodiff::Vector3dual u)
{
    autodiff::Vector4dual p1 = polynomial3d_1st_order(X);
    return u * p1.dot(a);
}

autodiff::dual energy(autodiff::Vector3dual X, autodiff::Vector3dual u, autodiff::Vector4dual a)
{
    using autodiff::at;
    using autodiff::gradient;
    using autodiff::jacobian;
    using autodiff::wrt;

    // IMPORTANT:
    // COMPUTING THE JACOBIAN MANUALLY WORKS FOR COMPUTING dPsi/du LATER ON.
    // COMMENT THIS SECTION TO DEBUG THE PROBLEM.
    autodiff::Vector3dual gradphi  = gradient(phi, wrt(X), at(X, a));
    autodiff::Matrix3dual Jacobian = u * gradphi.transpose();

    // IMPORTANT:
    // COMPUTING THE JACOBIAN WITH AUTODIFF DOESN'T WORK FOR COMPUTING dPsi/du LATER ON
    // UNCOMMENT THIS SECTION TO DEBUG THIS PART.
    // autodiff::Matrix3dual Jacobian = jacobian(interpolate, wrt(X), at(a, X, u));

    autodiff::dual Energy = Jacobian.squaredNorm();
    return Energy;
}

int main()
{
    autodiff::Vector4dual a;
    a << 1., 1., 1., 1.;
    autodiff::Vector3dual X;
    X << 1., 0., 0.;
    autodiff::Vector3dual u;
    u << 2., 0., 0.;

    using autodiff::at;
    using autodiff::gradient;
    using autodiff::wrt;

    autodiff::dual Psi;
    // Computing the gradient w.r.t. u here, but inside the energy function, 
    // we have to compute a derivative w.r.t. X
    autodiff::Vector3dual f = -gradient(energy, wrt(u), at(X, u, a), Psi);

    std::cout << "Psi:\n" << Psi << "\n";
    std::cout << "f:\n" << f << "\n";

    return 0;
}
```
Many thanks again for the support.

**Comment by allanleal** [comment]
OK, found some time to test your code. Replace

~~~
autodiff::Matrix3dual Jacobian = jacobian(interpolate, wrt(X), at(a, X, u));
~~~

by this

~~~c++
autodiff::Vector3dual ucopy = u;
autodiff::unseed(ucopy[0]);
autodiff::unseed(ucopy[1]);
autodiff::unseed(ucopy[2]);
autodiff::Matrix3dual Jacobian = jacobian(interpolate, wrt(X), at(a, X, ucopy));
~~~

The above tells the following. The seeded components in `u` when computing 

~~~c++
autodiff::Vector3dual f = -gradient(energy, wrt(u), at(X, u, a), Psi);
~~~

end up affecting the Jacobian computation

~~~
jacobian(interpolate, wrt(X), at(a, X, u));
~~~

By replacing `u` by a copy, with all `grad` members of `u[i]` unseeded, we avoid that effect.

Let me know if this clarifies. 

I would suggest not to use `autodiff::gradient/jacobian` one inside of the other, because this may result in unexpected behavior, unless you know what you are doing and can remove the unexpected effects.

**Comment by Q-Minh** [comment]
Amazing, this solves my problem! Thank you so much for looking into this problem. I'll be sure to remember how to seed forward passes correctly in the future when having nested derivative computations. Is it also inadvisable to use `autodiff::gradient` inside `autodiff::gradient` as I am doing currently in the lines
```
autodiff::Vector3dual gradphi  = gradient(phi, wrt(X), at(X, a));
autodiff::Matrix3dual Jacobian = u * gradphi.transpose();
```
of the `energy` function even when the differentiation variable is different (one gradient w.r.t. `X` nested in a gradient w.r.t. `u`)?

**Comment by allanleal** [comment]
Correct observation. If `u` was involved in this call `gradient(phi, wrt(X), at(X, a))`, say, something like this `gradient(phi, wrt(X), at(X, a, u))`, and `u` is indeed used inside of `phi`, then seeded variables in `u` would affect the gradient computation wrt `X`.

**Comment by tombelv** [comment]
Hi @allanleal , I hope it's ok if I post this here since it seems to me that my problem is related to this issue.
I am trying to compute the jacobian of a function `bar` of (x,p) with respect to a vector p, and inside this function I have to compute the jacobian of a function `foo` of (x,p) wrt x.

I can't manage to get things working even trying to unseed like you suggested for the problem above, so I came up with this simple example that shows the problem.

Do you have any clue on what I could do?

```
#include <autodiff/forward/dual.hpp>
#include <autodiff/forward/dual/eigen.hpp>
#include <iostream>

autodiff::VectorXdual foo(autodiff::VectorXdual& x, autodiff::VectorXdual& p) {
  // v = [x0        ]
  //     [p0*x1 + p1]
  autodiff::VectorXdual v(2);
  v(0) = x(0);
  v(1) = p(0)*x(1) +  p(1);
  return v;
}

autodiff::VectorXdual bar(autodiff::VectorXdual &x, autodiff::VectorXdual &p) {
  // This is the "trivial" implementation, does not work
  autodiff::MatrixXdual M = autodiff::jacobian(foo, autodiff::wrt(x),autodiff::at(x,p));
  // In theory:
  // M = [1   0]
  //     [0  p0]
  autodiff::VectorXdual res = M*p;
  return res;
}

autodiff::VectorXdual bar2(autodiff::VectorXdual &x, autodiff::VectorXdual &p) {
  autodiff::VectorXdual pcopy = p;
  for (int i = 0; i<pcopy.size(); ++i) autodiff::unseed(pcopy(i));
  autodiff::MatrixXdual M = autodiff::jacobian(foo, autodiff::wrt(x),autodiff::at(x,pcopy));
  // In this way it seems that we are bypassing the dependence of M on p

  autodiff::VectorXdual res = M*p;
  return res;
}



int main() {
  autodiff::VectorXdual x(2);
  x << 1,2;
  autodiff::VectorXdual p(2);
  p << 3,4;

  Eigen::MatrixXd Jp = autodiff::jacobian(bar, wrt(p), at(x, p));
  Eigen::MatrixXd Jp2 = autodiff::jacobian(bar2, wrt(p), at(x, p));

  std::cout << "M = \n" << autodiff::jacobian(foo, autodiff::wrt(x),autodiff::at(x,p)) << std::endl;
  // correctly prints M = [1 0; 0 p0] = [1 0; 0 3]

  std::cout << "Jp = \n" << Jp << std::endl;
  // Should print [1 0; p1, p0] = [1 0; 4 3]
  // but instead prints [1 0; 2 4]
  std::cout << "Jp2 = \n" << Jp2 << std::endl;
  // Should print [1 0; p1, p0] = [1 0; 4 3]
  // but instead prints [1 0; 0 3]

  return 0;
}

```

**Comment by allanleal** [comment]
Hi @tombelv , I just had a quick look at you code. The issue is that you are "consuming" autodiff already inside of your `bar function:

~~~c++
autodiff::VectorXdual bar(autodiff::VectorXdual &x, autodiff::VectorXdual &p) {
  // This is the "trivial" implementation, does not work
  autodiff::MatrixXdual M = autodiff::jacobian(foo, autodiff::wrt(x),autodiff::at(x,p));
  // In theory:
  // M = [1   0]
  //     [0  p0]
  autodiff::VectorXdual res = M*p;
  return res;
}
~~~

The matrix `M` produced is of type `MatrixXd` and therefore with `double` values, which can be considered `dual` numbers of order zero. After this point, you have no mechanism to propagate further the derivatives. 

I hope this clarifies. If that's not exactly the problem, apologies. Despite using `autodiff` heavily for my other libraries (e.g., https://reaktoro.org), and with plans to keep improving/extending `autodiff`, especially its forward algorithms, I currently find myself with limited time to provide technical support here. Thank you for your understanding and I hope the explanation above clarifies your issue.

**Comment by tombelv** [comment]
Thank you @allanleal for the quick answer, I think that you are right in saying that being the output of `jacobian` a `MatrixXd` the differentiation is not propagated.

I understand that you do not have much time so I apologize if asking another question is too much, but just in case: 
Does this mean that it is impossible in autodiff to "concatenate" jacobian calculations? Is there a workaround?

Thank you again.

**Comment by allanleal** [comment]
Hi @tombelv , you'd need to increase the order of your dual numbers (say, to second order). Then your `jacobian` call inside `bar` produces a `MatrixXdual1st` instead of `MatrixXdual0th === MatrixXd`. This way, you still have the chance to perform another round of derivative calculations. However, I have not tried this with your code. Your `bar` function would be declared like this:

~~~c++
autodiff::VectorXdual1st bar(autodiff::VectorXdual2nd &x, autodiff::VectorXdual2nd &p)
~~~

Note `VectorXdual` is alias for `VectorXdual1st`.

I think some users have tried to do this in the past (search for "nest derivative calculations" in the issues/discussions). This is something I don't do in my use case of `autodiff` so I'm not aware of how successfully people have been doing this.

**Comment by tombelv** [comment]
Unfortunately, I should have mentioned that I had already tried increasing the order of the dual variables, but to no effect.
I had also looked to past issues to see how people with my same problem had managed to solve it, but this seemed the most promising one.

I suspect that much further investigation would be needed to solve this, and I understand that this is not your concern at the moment, so thank you again for the help you have provided.
